From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Stained Glass OS <ke7oxh@gmail.com>
Date: Mon, 12 Oct 2026 21:00:08 -0700
Subject: [PATCH] ieframe: the browser's window object forwards to the page
 window; connection point enumerators

The window handed out for SID_SHTMLWindow (IHTMLWindow2 of IWebBrowser2) did
nothing for most of its methods; they now go to the window of the loaded page.
IConnectionPointContainer::EnumConnectionPoints is an enumerator over the
three connection points, and the connection enumerator gets Skip, Reset and
Clone, hands out referenced sinks and keeps the advise cookies.

diff --git a/dlls/ieframe/events.c b/dlls/ieframe/events.c
index 9c0711d..00b2a35 100644
--- a/dlls/ieframe/events.c
+++ b/dlls/ieframe/events.c
@@ -69,12 +69,175 @@ static ULONG WINAPI ConnectionPointContainer_Release(IConnectionPointContainer *
     return IUnknown_Release(This->impl);
 }
 
+typedef struct {
+    IEnumConnectionPoints IEnumConnectionPoints_iface;
+
+    LONG ref;
+    ULONG iter;
+    ConnectionPointContainer *container;
+} EnumConnectionPoints;
+
+static inline EnumConnectionPoints *impl_from_IEnumConnectionPoints(IEnumConnectionPoints *iface)
+{
+    return CONTAINING_RECORD(iface, EnumConnectionPoints, IEnumConnectionPoints_iface);
+}
+
+static HRESULT WINAPI EnumConnectionPoints_QueryInterface(IEnumConnectionPoints *iface, REFIID riid, void **ppv)
+{
+    EnumConnectionPoints *This = impl_from_IEnumConnectionPoints(iface);
+
+    if(IsEqualGUID(&IID_IUnknown, riid) || IsEqualGUID(&IID_IEnumConnectionPoints, riid)) {
+        TRACE("(%p)->(%s %p)\n", This, debugstr_guid(riid), ppv);
+        *ppv = &This->IEnumConnectionPoints_iface;
+        IUnknown_AddRef((IUnknown*)*ppv);
+        return S_OK;
+    }
+
+    WARN("Unsupported interface %s\n", debugstr_guid(riid));
+    *ppv = NULL;
+    return E_NOINTERFACE;
+}
+
+static ULONG WINAPI EnumConnectionPoints_AddRef(IEnumConnectionPoints *iface)
+{
+    EnumConnectionPoints *This = impl_from_IEnumConnectionPoints(iface);
+    LONG ref = InterlockedIncrement(&This->ref);
+
+    TRACE("(%p) ref=%ld\n", This, ref);
+
+    return ref;
+}
+
+static ULONG WINAPI EnumConnectionPoints_Release(IEnumConnectionPoints *iface)
+{
+    EnumConnectionPoints *This = impl_from_IEnumConnectionPoints(iface);
+    LONG ref = InterlockedDecrement(&This->ref);
+
+    TRACE("(%p) ref=%ld\n", This, ref);
+
+    if(!ref) {
+        IUnknown_Release(This->container->impl);
+        free(This);
+    }
+    return ref;
+}
+
+static ConnectionPoint *enum_cp_get(EnumConnectionPoints *This, ULONG index)
+{
+    switch(index) {
+    case 0: return This->container->wbe2;
+    case 1: return This->container->wbe;
+    case 2: return This->container->pns;
+    }
+    return NULL;
+}
+
+static HRESULT WINAPI EnumConnectionPoints_Next(IEnumConnectionPoints *iface, ULONG cConnections,
+        IConnectionPoint **ppCP, ULONG *pcFetched)
+{
+    EnumConnectionPoints *This = impl_from_IEnumConnectionPoints(iface);
+    ULONG cnt = 0;
+
+    TRACE("(%p)->(%lu %p %p)\n", This, cConnections, ppCP, pcFetched);
+
+    if(!ppCP || (!pcFetched && cConnections > 1))
+        return E_POINTER;
+
+    while(cnt < cConnections) {
+        ConnectionPoint *cp = enum_cp_get(This, This->iter);
+        if(!cp)
+            break;
+        ppCP[cnt] = &cp->IConnectionPoint_iface;
+        IConnectionPoint_AddRef(ppCP[cnt]);
+        This->iter++;
+        cnt++;
+    }
+
+    if(pcFetched)
+        *pcFetched = cnt;
+    return cnt == cConnections ? S_OK : S_FALSE;
+}
+
+static HRESULT WINAPI EnumConnectionPoints_Skip(IEnumConnectionPoints *iface, ULONG cConnections)
+{
+    EnumConnectionPoints *This = impl_from_IEnumConnectionPoints(iface);
+
+    TRACE("(%p)->(%lu)\n", This, cConnections);
+
+    while(cConnections--) {
+        if(!enum_cp_get(This, This->iter))
+            return S_FALSE;
+        This->iter++;
+    }
+    return S_OK;
+}
+
+static HRESULT WINAPI EnumConnectionPoints_Reset(IEnumConnectionPoints *iface)
+{
+    EnumConnectionPoints *This = impl_from_IEnumConnectionPoints(iface);
+
+    TRACE("(%p)\n", This);
+
+    This->iter = 0;
+    return S_OK;
+}
+
+static HRESULT WINAPI EnumConnectionPoints_Clone(IEnumConnectionPoints *iface, IEnumConnectionPoints **ppEnum)
+{
+    EnumConnectionPoints *This = impl_from_IEnumConnectionPoints(iface);
+    EnumConnectionPoints *ret;
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+    if(!ppEnum)
+        return E_POINTER;
+    if(!(ret = malloc(sizeof(*ret))))
+        return E_OUTOFMEMORY;
+
+    ret->IEnumConnectionPoints_iface.lpVtbl = This->IEnumConnectionPoints_iface.lpVtbl;
+    ret->ref = 1;
+    ret->iter = This->iter;
+    ret->container = This->container;
+    IUnknown_AddRef(ret->container->impl);
+
+    *ppEnum = &ret->IEnumConnectionPoints_iface;
+    return S_OK;
+}
+
+static const IEnumConnectionPointsVtbl EnumConnectionPointsVtbl = {
+    EnumConnectionPoints_QueryInterface,
+    EnumConnectionPoints_AddRef,
+    EnumConnectionPoints_Release,
+    EnumConnectionPoints_Next,
+    EnumConnectionPoints_Skip,
+    EnumConnectionPoints_Reset,
+    EnumConnectionPoints_Clone
+};
+
 static HRESULT WINAPI ConnectionPointContainer_EnumConnectionPoints(IConnectionPointContainer *iface,
         LPENUMCONNECTIONPOINTS *ppEnum)
 {
     ConnectionPointContainer *This = impl_from_IConnectionPointContainer(iface);
-    FIXME("(%p)->(%p)\n", This, ppEnum);
+    EnumConnectionPoints *ret;
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+#ifdef SG_MUTANT_IEFRAME_NOCPENUM
     return E_NOTIMPL;
+#endif
+    if(!ppEnum)
+        return E_POINTER;
+    if(!(ret = malloc(sizeof(*ret))))
+        return E_OUTOFMEMORY;
+
+    ret->IEnumConnectionPoints_iface.lpVtbl = &EnumConnectionPointsVtbl;
+    ret->ref = 1;
+    ret->iter = 0;
+    ret->container = This;
+    IUnknown_AddRef(This->impl);
+
+    *ppEnum = &ret->IEnumConnectionPoints_iface;
+    return S_OK;
 }
 
 static HRESULT WINAPI ConnectionPointContainer_FindConnectionPoint(IConnectionPointContainer *iface,
@@ -201,7 +364,8 @@ static HRESULT WINAPI EnumConnections_Next(IEnumConnections *iface, ULONG cConne
             break;
 
         pgcd[cnt].pUnk = (IUnknown*)This->cp->sinks[This->iter];
-        pgcd[cnt].dwCookie = cnt+1;
+        IUnknown_AddRef(pgcd[cnt].pUnk);
+        pgcd[cnt].dwCookie = This->iter+1;
         This->iter++;
         cnt++;
     }
@@ -214,22 +378,48 @@ static HRESULT WINAPI EnumConnections_Next(IEnumConnections *iface, ULONG cConne
 static HRESULT WINAPI EnumConnections_Skip(IEnumConnections *iface, ULONG cConnections)
 {
     EnumConnections *This = impl_from_IEnumConnections(iface);
-    FIXME("(%p)->(%lu)\n", This, cConnections);
-    return E_NOTIMPL;
+
+    TRACE("(%p)->(%lu)\n", This, cConnections);
+
+    while(cConnections) {
+        while(This->iter < This->cp->sinks_size && !This->cp->sinks[This->iter])
+            This->iter++;
+        if(This->iter == This->cp->sinks_size)
+            return S_FALSE;
+        This->iter++;
+        cConnections--;
+    }
+    return S_OK;
 }
 
 static HRESULT WINAPI EnumConnections_Reset(IEnumConnections *iface)
 {
     EnumConnections *This = impl_from_IEnumConnections(iface);
-    FIXME("(%p)\n", This);
-    return E_NOTIMPL;
+
+    TRACE("(%p)\n", This);
+
+    This->iter = 0;
+    return S_OK;
 }
 
 static HRESULT WINAPI EnumConnections_Clone(IEnumConnections *iface, IEnumConnections **ppEnum)
 {
     EnumConnections *This = impl_from_IEnumConnections(iface);
-    FIXME("(%p)->(%p)\n", This, ppEnum);
-    return E_NOTIMPL;
+    EnumConnections *ret;
+
+    TRACE("(%p)->(%p)\n", This, ppEnum);
+
+    if(!(ret = malloc(sizeof(*ret))))
+        return E_OUTOFMEMORY;
+
+    ret->IEnumConnections_iface.lpVtbl = This->IEnumConnections_iface.lpVtbl;
+    ret->ref = 1;
+    ret->iter = This->iter;
+    ret->cp = This->cp;
+    IConnectionPoint_AddRef(&ret->cp->IConnectionPoint_iface);
+
+    *ppEnum = &ret->IEnumConnections_iface;
+    return S_OK;
 }
 
 static const IEnumConnectionsVtbl EnumConnectionsVtbl = {
diff --git a/dlls/ieframe/iehtmlwnd.c b/dlls/ieframe/iehtmlwnd.c
index de211fc..f58352e 100644
--- a/dlls/ieframe/iehtmlwnd.c
+++ b/dlls/ieframe/iehtmlwnd.c
@@ -93,16 +93,40 @@ static ULONG WINAPI IEHTMLWindow2_Release(IHTMLWindow2 *iface)
 static HRESULT WINAPI IEHTMLWindow2_GetTypeInfoCount(IHTMLWindow2 *iface, UINT *pctinfo)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_GetTypeInfoCount(window, pctinfo);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_GetTypeInfo(IHTMLWindow2 *iface, UINT iTInfo,
                                               LCID lcid, ITypeInfo **ppTInfo)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_GetTypeInfo(window, iTInfo, lcid, ppTInfo);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_GetIDsOfNames(IHTMLWindow2 *iface, REFIID riid,
@@ -110,8 +134,20 @@ static HRESULT WINAPI IEHTMLWindow2_GetIDsOfNames(IHTMLWindow2 *iface, REFIID ri
                                                 LCID lcid, DISPID *rgDispId)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_GetIDsOfNames(window, riid, rgszNames, cNames, lcid, rgDispId);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_Invoke(IHTMLWindow2 *iface, DISPID dispIdMember,
@@ -119,116 +155,308 @@ static HRESULT WINAPI IEHTMLWindow2_Invoke(IHTMLWindow2 *iface, DISPID dispIdMem
                             VARIANT *pVarResult, EXCEPINFO *pExcepInfo, UINT *puArgErr)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_Invoke(window, dispIdMember, riid, lcid, wFlags, pDispParams, pVarResult, pExcepInfo, puArgErr);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_item(IHTMLWindow2 *iface, VARIANT *pvarIndex, VARIANT *pvarResult)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %p)\n", This, debugstr_variant(pvarIndex), pvarResult);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_item(window, pvarIndex, pvarResult);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_length(IHTMLWindow2 *iface, LONG *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_length(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_frames(IHTMLWindow2 *iface, IHTMLFramesCollection2 **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_frames(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_defaultStatus(IHTMLWindow2 *iface, BSTR v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_defaultStatus(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_defaultStatus(IHTMLWindow2 *iface, BSTR *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_defaultStatus(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_status(IHTMLWindow2 *iface, BSTR v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_status(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_status(IHTMLWindow2 *iface, BSTR *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_status(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_setTimeout(IHTMLWindow2 *iface, BSTR expression,
         LONG msec, VARIANT *language, LONG *timerID)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %ld %s %p)\n", This, debugstr_w(expression), msec, debugstr_variant(language), timerID);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_setTimeout(window, expression, msec, language, timerID);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_clearTimeout(IHTMLWindow2 *iface, LONG timerID)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld)\n", This, timerID);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_clearTimeout(window, timerID);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_alert(IHTMLWindow2 *iface, BSTR message)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(message));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_alert(window, message);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_confirm(IHTMLWindow2 *iface, BSTR message,
         VARIANT_BOOL *confirmed)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %p)\n", This, debugstr_w(message), confirmed);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_confirm(window, message, confirmed);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_prompt(IHTMLWindow2 *iface, BSTR message,
         BSTR dststr, VARIANT *textdata)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %s %p)\n", This, debugstr_w(message), debugstr_w(dststr), textdata);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_prompt(window, message, dststr, textdata);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_Image(IHTMLWindow2 *iface, IHTMLImageElementFactory **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_Image(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_location(IHTMLWindow2 *iface, IHTMLLocation **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_location(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_history(IHTMLWindow2 *iface, IOmHistory **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_history(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_close(IHTMLWindow2 *iface)
@@ -246,206 +474,558 @@ static HRESULT WINAPI IEHTMLWindow2_close(IHTMLWindow2 *iface)
 static HRESULT WINAPI IEHTMLWindow2_put_opener(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_opener(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_opener(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_opener(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_navigator(IHTMLWindow2 *iface, IOmNavigator **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
-}
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_navigator(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
+}
 
 static HRESULT WINAPI IEHTMLWindow2_put_name(IHTMLWindow2 *iface, BSTR v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(v));
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    
+#ifdef SG_MUTANT_IEFRAME_NOFWD
+    IHTMLWindow2_Release(window);
     return E_NOTIMPL;
+#endif
+    hres = IHTMLWindow2_put_name(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_name(IHTMLWindow2 *iface, BSTR *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_name(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_parent(IHTMLWindow2 *iface, IHTMLWindow2 **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_parent(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_open(IHTMLWindow2 *iface, BSTR url, BSTR name,
          BSTR features, VARIANT_BOOL replace, IHTMLWindow2 **pomWindowResult)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %s %s %x %p)\n", This, debugstr_w(url), debugstr_w(name),
-          debugstr_w(features), replace, pomWindowResult);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_open(window, url, name, features, replace, pomWindowResult);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_self(IHTMLWindow2 *iface, IHTMLWindow2 **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_self(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_top(IHTMLWindow2 *iface, IHTMLWindow2 **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_top(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_window(IHTMLWindow2 *iface, IHTMLWindow2 **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_window(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_navigate(IHTMLWindow2 *iface, BSTR url)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_w(url));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_navigate(window, url);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onfocus(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onfocus(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onfocus(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onfocus(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onblur(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onblur(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onblur(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onblur(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onload(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onload(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onload(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onload(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onbeforeunload(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onbeforeunload(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onbeforeunload(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onbeforeunload(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onunload(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onunload(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onunload(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onunload(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onhelp(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onhelp(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onhelp(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onhelp(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onerror(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onerror(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onerror(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onerror(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onresize(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onresize(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onresize(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onresize(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_onscroll(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_onscroll(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_onscroll(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_onscroll(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_document(IHTMLWindow2 *iface, IHTMLDocument2 **p)
@@ -472,174 +1052,461 @@ static HRESULT WINAPI IEHTMLWindow2_get_document(IHTMLWindow2 *iface, IHTMLDocum
 static HRESULT WINAPI IEHTMLWindow2_get_event(IHTMLWindow2 *iface, IHTMLEventObj **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_event(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get__newEnum(IHTMLWindow2 *iface, IUnknown **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get__newEnum(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_showModalDialog(IHTMLWindow2 *iface, BSTR dialog,
         VARIANT *varArgIn, VARIANT *varOptions, VARIANT *varArgOut)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %s %s %p)\n", This, debugstr_w(dialog), debugstr_variant(varArgIn),
-        debugstr_variant(varOptions), varArgOut);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_showModalDialog(window, dialog, varArgIn, varOptions, varArgOut);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_showHelp(IHTMLWindow2 *iface, BSTR helpURL, VARIANT helpArg,
         BSTR features)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %s %s)\n", This, debugstr_w(helpURL), debugstr_variant(&helpArg), debugstr_w(features));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_showHelp(window, helpURL, helpArg, features);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_screen(IHTMLWindow2 *iface, IHTMLScreen **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_screen(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_Option(IHTMLWindow2 *iface, IHTMLOptionElementFactory **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_Option(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_focus(IHTMLWindow2 *iface)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_focus(window);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_closed(IHTMLWindow2 *iface, VARIANT_BOOL *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_closed(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_blur(IHTMLWindow2 *iface)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->()\n", This);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_blur(window);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_scroll(IHTMLWindow2 *iface, LONG x, LONG y)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld %ld)\n", This, x, y);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_scroll(window, x, y);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_clientInformation(IHTMLWindow2 *iface, IOmNavigator **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_clientInformation(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_setInterval(IHTMLWindow2 *iface, BSTR expression,
         LONG msec, VARIANT *language, LONG *timerID)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %ld %s %p)\n", This, debugstr_w(expression), msec, debugstr_variant(language), timerID);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_setInterval(window, expression, msec, language, timerID);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_clearInterval(IHTMLWindow2 *iface, LONG timerID)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld)\n", This, timerID);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_clearInterval(window, timerID);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_put_offscreenBuffering(IHTMLWindow2 *iface, VARIANT v)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s)\n", This, debugstr_variant(&v));
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_put_offscreenBuffering(window, v);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_offscreenBuffering(IHTMLWindow2 *iface, VARIANT *p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_offscreenBuffering(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_execScript(IHTMLWindow2 *iface, BSTR scode, BSTR language,
         VARIANT *pvarRet)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%s %s %p)\n", This, debugstr_w(scode), debugstr_w(language), pvarRet);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_execScript(window, scode, language, pvarRet);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_toString(IHTMLWindow2 *iface, BSTR *String)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, String);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_toString(window, String);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_scrollBy(IHTMLWindow2 *iface, LONG x, LONG y)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld %ld)\n", This, x, y);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_scrollBy(window, x, y);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_scrollTo(IHTMLWindow2 *iface, LONG x, LONG y)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld %ld)\n", This, x, y);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_scrollTo(window, x, y);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_moveTo(IHTMLWindow2 *iface, LONG x, LONG y)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld %ld)\n", This, x, y);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_moveTo(window, x, y);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_moveBy(IHTMLWindow2 *iface, LONG x, LONG y)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld %ld)\n", This, x, y);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_moveBy(window, x, y);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_resizeTo(IHTMLWindow2 *iface, LONG x, LONG y)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld %ld)\n", This, x, y);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_resizeTo(window, x, y);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_resizeBy(IHTMLWindow2 *iface, LONG x, LONG y)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%ld %ld)\n", This, x, y);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_resizeBy(window, x, y);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static HRESULT WINAPI IEHTMLWindow2_get_external(IHTMLWindow2 *iface, IDispatch **p)
 {
     IEHTMLWindow *This = impl_from_IHTMLWindow2(iface);
-    FIXME("(%p)->(%p)\n", This, p);
-    return E_NOTIMPL;
+    IHTMLWindow2 *window;
+    HRESULT hres;
+
+    TRACE("(%p)\n", This);
+
+    hres = get_window(This->doc_host, &window);
+    if(FAILED(hres))
+        return hres;
+    if(!window)
+        return E_FAIL;
+
+    hres = IHTMLWindow2_get_external(window, p);
+    IHTMLWindow2_Release(window);
+    return hres;
 }
 
 static const IHTMLWindow2Vtbl IEHTMLWindow2Vtbl = {
